### PR TITLE
[QHC-777] Major refactor to simplify things

### PR DESCRIPTION
**Pull request by fedonman**
**Major reorganization of the library structure and runcard functionality**. Key updates include:
  - Removed obsolete instruments, such as VNAs.
  - Removed the `drivers` module.
  - Simplified the `Qblox` sequencer class hierarchy into two main classes: `QbloxSequencer` and `QbloxADCSequencer`.
  - Removed `SystemController` and `ReadoutSystemController`; buses now interface directly with instruments.
  - Introduced a new `channels` attribute to the `Bus` class, allowing specification of channels for each associated instrument.
  - Removed the `Chip` class and its related runcard settings.
  - Eliminated outdated settings, such as instrument firmware.
  - Refactored runcard settings into a modular structure with four distinct groups:
    - `instruments` and `instrument_controllers` for lab instrument setup.
    - `buses` for grouping instrument channels.
    - `digital` for digital compilation settings (e.g., Qibo circuits).
    - `analog` for analog compilation settings (e.g., annealing programs).

### CODE REVIEW

File: src/qililab/instruments/qblox/qblox_qcm.py [hunk 1]
```diff
@@ -35,3 +36,11 @@ class QbloxQCMSettings(QbloxModule.QbloxModuleSettings):
...
 
     settings: QbloxQCMSettings
+
+    def is_awg(self) -> bool:
+        """Returns True if instrument is an AWG."""
+        return True
+
+    def is_adc(self) -> bool:
+        """Returns True if instrument is an ADC."""
+        return False
```
Comment (jordivallsq) [inline src/qililab/instruments/qblox/qblox_qcm.py:46]: I don't understand why in some cases you have False here despite its description

File: src/qililab/instruments/qblox/qblox_qrm.py [hunk 2]
```diff
@@ -41,35 +40,38 @@ class QbloxQRM(QbloxModule, AWGAnalogDigitalConverter):
...
     settings: QbloxQRMSettings
 
-    @Instrument.CheckDeviceInitialized
+    def is_awg(self) -> bool:
+        """Returns True if instrument is an AWG."""
+        return True
+
+    def is_adc(self) -> bool:
+        """Returns True if instrument is an AWG/ADC."""
+        return True
```
Comment (jordivallsq) [inline src/qililab/instruments/qblox/qblox_qrm.py:72]: Here on the other hand it is set to True

File: src/qililab/instruments/qblox/qblox_qcm.py (hunk 1 above)
Comment (fedonman) [reply src/qililab/instruments/qblox/qblox_qcm.py:46]: `is_awg` is True if the instrument contains an Arbitrary Waveform Generator, so in the case of QCM it is True.
`is_adc` is True if the instrument contains an Analog to Digital Converter, so in the case of QCM it is False.

The default of both in the parent class `Instrument` is False. So actually I can remove this overload.

Comment (jordivallsq) [reply src/qililab/instruments/qblox/qblox_qcm.py]: Ok now I understand, I misunderstood the functions and the documentation thinking it worked the other way around (you set the is_....() and then it updates your machine info), now it makes sense.

File: tests/result/test_qblox_result.py [hunk 3]
```diff
@@ -1,377 +1,377 @@
...
-        with pytest.raises(
-            NotImplementedError, match="Samples for multiple measurements on a single qubit are not supported"
-        ):
-            _ = qblox_multi_m_results.samples()
-
-    def test_to_dataframe(self, qblox_result_noscope: QbloxResult):
-        """Test the to_dataframe method."""
-        dataframe = qblox_result_noscope.to_dataframe()
-        assert isinstance(dataframe, pd.DataFrame)
+# """ Test Results """
```
Comment (jordivallsq) [inline tests/result/test_qblox_result.py:1]: Why is this file commented instead of deleted?

Comment (fedonman) [reply tests/result/test_qblox_result.py:1]: Nice catch, I deleted it.

File: src/qililab/digital/circuit_transpiler.py [hunk 4]
```diff
@@ -337,14 +324,14 @@ def _gate_element_to_pulse_event(
...
-        # for couplers we don't need to set the target qubit
-        elif isinstance(bus.targets[0], Coupler):
-            qubit = None
-        # handle control gates, target should be the qubit target of the bus
-        else:
-            qubit = next(target.qubit_index for target in bus.targets if isinstance(target, Qubit))
+        bus = self.digital_compilation_settings.buses[gate_event.bus]
+        qubit = (
+            gate.qubits[0]
+            if isinstance(gate, M)
```
Comment (jordivallsq) [inline src/qililab/digital/circuit_transpiler.py:330]: I get that this M gate class is comming from qibo but can we do something regarding that? It is not ok to use a class that is a single letter, I propose that at least we change the name of such imported variables like with from qibo.gates import M as Measure_gate or something

File: src/qililab/digital/circuit_transpiler.py [hunk 5]
```diff
@@ -110,7 +110,9 @@ def optimize_transpilation(self, nqubits: int, ngates: list[gates.Gate]) -> list
                 shift[gate.qubits[0]] += gate.parameters[0]
             # add CZ phase correction
             elif isinstance(gate, gates.CZ):
```
Comment (jordivallsq) [inline src/qililab/digital/circuit_transpiler.py:112]: We are also using gate for CZ instead of using from qibo.gates import ... (again I know that this is not related to your changes but I would like to point that out)

File: tests/digital/test_circuit_transpiler.py [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+    ]
+
+@pytest.fixture(name="digital_settings")
+def fixture_digital_compilation_settings() -> DigitalCompilationSettings:
+    """Fixture that returns an instance of a ``Runcard.GatesSettings`` class."""
+    digital_settings_dict = {
+        "minimum_clock_time": 5,
+        "delay_before_readout": 0,
+        "topology": [(0, 2), (1, 2), (2, 3), (2, 4)],
+        "gates": {
```
[the remark this answers is not in the document]
Comment (jordivallsq) [reply tests/digital/test_circuit_transpiler.py:217]: based on the structure that I saw in tests/calibration/galadriel.yml

File: .github/codecov.yml [hunk 7]
```diff
@@ -33,3 +32,8 @@ coverage:
           - "unittests"
         paths:
           - "src"
+
+ignore:
+  - "src/qililab/instruments/keithley/keithley_2600.py"
```
[the remark this answers is not in the document]
Comment (jordivallsq) [reply .github/codecov.yml:37]: The same happens for qblox_d5a as the tests are also modified, I see no superficial problem in those tests. For qblox_s4g it makes more sense as the tests `TestQbloxS4g` are empty.

What I don't understand is the empty of turn on / off for keithley.

[the remark this answers is not in the document]
Comment (fedonman) [reply .github/codecov.yml:37]: there are partial or no tests for these instruments, so I don't want codecov to count them towards 100% for this PR.

File: pyproject.toml [hunk 8]
```diff
@@ -140,7 +142,11 @@ tag_format = "$version"
         "*/virtualenv/*",
         "*/virtualenvs/*",
         "*/tests/*",
-        "examples/*"
+        "examples/*",
+        # TODO: Remove these once we modernize instruments
+        "src/qililab/instruments/keithley/keithley_2600.py",
```
[the remark this answers is not in the document]
Comment (fedonman) [reply pyproject.toml:147]: yes

File: tests/digital/test_circuit_transpiler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fedonman) [reply tests/digital/test_circuit_transpiler.py:217]: I didn't understand the question.

File: src/qililab/digital/circuit_transpiler.py (hunk 4 above)
Comment (fedonman) [reply src/qililab/digital/circuit_transpiler.py]: I changed the import to `gates.M`.

File: tests/digital/test_circuit_transpiler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fedonman) [reply tests/digital/test_circuit_transpiler.py:217]: digital has `gates` included.

image

File: src/qililab/platform/components/bus.py [hunk 9]
```diff
@@ -197,49 +180,78 @@ def get_parameter(self, parameter: Parameter, channel_id: int | None = None):
...
+        for instrument, instrument_channel in zip(self.instruments, self.channels):
+            if isinstance(instrument, (QbloxQCM, QbloxQRM)):
+                instrument.upload(channel_id=instrument_channel)
+                return
 
     def run(self) -> None:
         """Runs any previously uploaded program into the instrument."""
-        self.system_control.run(port=self.port)
+        for instrument, instrument_channel in zip(self.instruments, self.channels):
+            if isinstance(instrument, (QbloxQCM, QbloxQRM)):
```
Comment (jordivallsq) [inline src/qililab/platform/components/bus.py:211]: Why here in bus we only use Qblox? I don't understand what happens with QM

Comment (fedonman) [reply src/qililab/platform/components/bus.py:211]: This is a method that is executed only when executing a qibo circuit which is supported in qblox only

File: src/qililab/platform/platform.py [hunk 10]
```diff
@@ -1105,16 +1032,29 @@ def compile(
...
-        return self.compiler.compile(
+        module_and_sequencer_per_bus: dict[str, ModuleSequencer] = {
+            element.bus_alias: ModuleSequencer(module=instrument, sequencer=instrument.get_sequencer(channel))
+            for element in pulse_schedule.elements
+            for instrument, channel in zip(
+                self.buses.get(alias=element.bus_alias).instruments, self.buses.get(alias=element.bus_alias).channels
+            )
+            if isinstance(instrument, QbloxModule)
+        }
+        compiler = PulseQbloxCompiler(
```
Comment (jordivallsq) [inline src/qililab/platform/platform.py:1054]: Here in this compile inside platform, can you confirm that this is Qblox "exclusive" because it is the pulse schedule?

Comment (fedonman) [reply src/qililab/platform/platform.py:1054]: Yes, anything that has to do with the `pulse` module, and transpilation from Qibo -> PulseSchedule -> Q1ASM is Qblox specific.

File: src/qililab/settings/settings.py [hunk 11]
```diff
@@ -29,81 +26,3 @@ class Settings:
     def __post_init__(self):
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/settings/settings.py:26]: I guess it is there because we want it to execute for all subclasses of Settings.

**Review by jordivallsq — APPROVED** [review]
Revised the code and tested the creation of platform using the new runcard. Great job!!
